catalyst/contrib/datasets/metric_learning.py

```diff
@@ -0,0 +1,146 @@
+from typing import Any, Callable, Dict, List, Optional
+from abc import ABC, abstractmethod
+
+import torch
+from torch.utils.data import Dataset
+
+from catalyst.contrib.datasets import MNIST
+
+
+class MetricLearningTrainDataset(Dataset, ABC):
+    """
+    Base class for datasets adapted for
+    metric learning train stage.
+    """
+
+    @abstractmethod
+    def get_labels(self) -> List[int]:
+        """
+        Dataset for metric learning must provide
+        label of each sample for forming positive
+        and negative pairs during the training
+        based on these labels.
+
+        Returns:
+            labels of samples
+        """
+        raise NotImplementedError
+
+
+class MnistMLDataset(MetricLearningTrainDataset, MNIST):
+    """
+    Simple wrapper for MNIST dataset
+    """
+
+    def get_labels(self) -> List[int]:
+        """
+        Returns:
+            labels of digits
+        """
+        return self.targets.tolist()
+
+
+class QueryGalleryDataset(Dataset, ABC):
+    """
+    QueryGallleryDataset for CMCScoreCallback
+    """
+
+    @abstractmethod
+    def __getitem__(self, item) -> Dict[str, torch.Tensor]:
+        """
+        Dataset for query/gallery split should
+        return dict with `feature`, `targets` and
+        `is_query` key. Value by key `is_query` should
+        be boolean and indicate whether current object
+        is in query or in gallery.
+
+        Raises:
+            NotImplementedError: You should implement it  # noqa: DAR402
+        """
+        raise NotImplementedError()
+
+    @property
+    @abstractmethod
+    def query_size(self) -> int:
+        """
+        Query/Gallery dataset should have property
+        query size.
+
+        Raises:
+            NotImplementedError: You should implement it  # noqa: DAR402
+        Returns:
+            query size
+        """
+        raise NotImplementedError()
+
+    @property
+    @abstractmethod
+    def gallery_size(self) -> int:
+        """
+        Query/Gallery dataset should have property
+        gallery size.
+
+        Raises:
+            NotImplementedError: You should implement it  # noqa: DAR402
+        Returns:
+            gallery size
+        """
+        raise NotImplementedError()
+
+
+class MnistQGDataset(QueryGalleryDataset):
```

Comment:
what do you think about merging all mnist datasets to `datasets/mnist.py` and metric-learning-based datasets to `datasets/metric_learning`?

Reply:
done